### PR TITLE
feat(react-scheduler): redesign AppointmentTooltip

### PR DESCRIPTION
**Pull request by AryamnovEugeniy**
Before:

image  image

After:

image image
image

BREAKING CHANGE:
We have extended the `headerComponent` and `contentComponent` in the `AppointmentTooltip` plugin with new properties. If you use these components, add the following properties in your code:

```diff
...
<AppointmentTooltip
  headerComponent={({
+  commandButtonComponent
+  showOpenButton
+  showCloseButton
+  showDeleteButton
+  commandBottonIds
     ...restProps
  }) => (
    <AppointmentTooltip.Header
+      commandButtonComponent={commandButtonComponent}
+      showOpenButton={showOpenButton}
+      showCloseButton={showCloseButton}
+      showDeleteButton={showDeleteButton}
+      commandBottonIds={commandBottonIds}
       {...restProps}
    />
  )}
contentComponent={({
+   formatDate
     ...restProps
  }) => (
    <AppointmentTooltip.Content
+      formatDate={formatDate}
       {...restProps}
    />
  )}
/>
...
```

### CODE REVIEW

**Comment by AryamnovEugeniy** [comment]
FPR

**Comment by MaximKudriavtsev** [comment]
Can we add a padding top (or something else)? I think 2px would enough
image

File: packages/dx-react-scheduler/src/types/editing/appointment-tooltip.types.ts [hunk 1]
```diff
@@ -33,13 +35,31 @@ export namespace AppointmentTooltip {
...
+    /** An event raised when the Open button is clicked. The event handler should open the appointment form. */
+    onOpenButtonClick?: () => void;
+    /** An event raised when the Open button is clicked. The event handler should delete an appointment. */
+    onDeleteButtonClick?: () => void;
+    /** An event that hides the tooltip. */
+    onHide?: () => void;
+    /** A component that renders a command button. */
+    commandButtonComponent: React.ComponentType<AppointmentTooltip.CommandButtonProps>;
     /** A React node used to render the tooltip header. */
     children?: React.ReactNode;
```
Comment (MaximKudriavtsev) [inline packages/dx-react-scheduler/src/types/editing/appointment-tooltip.types.ts]: `A React node used to render the additional part into tooltip header.`

**Comment by AryamnovEugeniy** [comment]
FPR

File: packages/dx-react-scheduler-material-ui/src/templates/appointment-tooltip/header.jsx [hunk 2]
```diff
@@ -3,44 +3,93 @@ import * as PropTypes from 'prop-types';
...
-    display: 'flex',
-    flexDirection: 'column',
+    width: '400px',
     position: 'relative',
-    paddingLeft: theme.spacing(1),
-    paddingRight: theme.spacing(1),
+    paddingLeft: spacing(1),
+    paddingRight: spacing(0.5),
+    paddingTop: spacing(0.25),
+    minHeight: spacing(3),
```
Comment (MaximKudriavtsev) [inline packages/dx-react-scheduler-material-ui/src/templates/appointment-tooltip/header.jsx]: `minHeight: spacing(3)` -> `minHeight: spacing(1.5),` // 12 px would be better

**Comment by AryamnovEugeniy** [comment]
FPR

**Comment by AryamnovEugeniy** [comment]
FPR

File: packages/dx-react-scheduler/docs/reference/appointment-tooltip.md [hunk 3]
```diff
@@ -40,6 +40,7 @@ layoutComponent | ComponentType&lt;[AppointmentTooltip.LayoutProps](#appointment
 headerComponent | ComponentType&lt;[AppointmentTooltip.HeaderProps](#appointmenttooltipheaderprops)&gt; | | A component that renders the header.
 contentComponent | ComponentType&lt;[AppointmentTooltip.ContentProps](#appointmenttooltipcontentprops)&gt; | | A component that renders the tooltip content.
 commandButtonComponent | ComponentType&lt;[AppointmentTooltip.CommandButtonProps](#appointmenttooltipcommandbuttonprops)&gt; | | A component that renders a command button.
+recurringIconComponent | ComponentType&lt;object&gt; | | A component that renders an icon for recurring appointments inside the AppointmentTooltip.
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/reference/appointment-tooltip.md]:
```suggestion
recurringIconComponent | ComponentType&lt;object&gt; | | A component that renders an icon that indicates a recurring appointment.
```

File: packages/dx-react-scheduler/docs/reference/appointment-tooltip.md [hunk 4]
```diff
@@ -64,22 +65,30 @@ showDeleteButton | boolean | Specifies the Delete button's visibility.
...
-onDeleteButtonClick? | () => void | An event raised when the Open button is clicked. The event handler should delete an appointment.
+onDeleteButtonClick? | () => void | An event raised when the Delete button is clicked. The event handler should delete an appointment.
 appointmentMeta? | [AppointmentMeta](#appointmentmeta) | The appointment's displayed metadata.
 visible? | boolean | Specifies the tooltip's visibility.
 onHide? | () => void | An event that hides the tooltip.
 headerComponent | ComponentType&lt;[AppointmentTooltip.HeaderProps](#appointmenttooltipheaderprops)&gt; | A component that renders the tooltip header.
 contentComponent | ComponentType&lt;[AppointmentTooltip.ContentProps](#appointmenttooltipcontentprops)&gt; | A component that renders the tooltip content.
 commandButtonComponent | ComponentType&lt;[AppointmentTooltip.CommandButtonProps](#appointmenttooltipcommandbuttonprops)&gt; | A component that renders a command button.
+recurringIconComponent | ComponentType&lt;object&gt; | A component that renders an icon for recurring appointments inside the AppointmentTooltip.
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/reference/appointment-tooltip.md]:
```suggestion
recurringIconComponent | ComponentType&lt;object&gt; | A component that renders an icon that indicates a recurring appointment.
```

File: packages/dx-react-scheduler/docs/reference/appointment-tooltip.md [hunk 5]
```diff
@@ -64,22 +65,30 @@ showDeleteButton | boolean | Specifies the Delete button's visibility.
...
 Field | Type | Description
 ------|------|------------
+showOpenButton? | boolean | Specifies the Open button's visibility.
+showCloseButton? | boolean | Specifies the Close button's visibility.
+showDeleteButton? | boolean | Specifies the Delete button's visibility.
 appointmentData? | [AppointmentModel](./scheduler.md#appointmentmodel) | The appointment's displayed metadata.
-children? | ReactNode | A React node used to render the tooltip header.
+onOpenButtonClick? | () => void | An event raised when the Open button is clicked. The event handler should open the appointment form.
+onDeleteButtonClick? | () => void | An event raised when the Delete button is clicked. The event handler should delete an appointment.
+onHide? | () => void | An event that hides the tooltip.
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/reference/appointment-tooltip.md]:
```suggestion
onHide? | () => void | An event raised when the tooltip hides.
```

File: packages/dx-react-scheduler/docs/reference/appointment-tooltip.md [hunk 6]
```diff
@@ -64,22 +65,30 @@ showDeleteButton | boolean | Specifies the Delete button's visibility.
...
+commandButtonComponent | ComponentType&lt;[AppointmentTooltip.CommandButtonProps](#appointmenttooltipcommandbuttonprops)&gt; | A component that renders a command button.
+children? | ReactNode | A React node used to render the additional part of the tooltip header.
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/reference/appointment-tooltip.md]:
```suggestion
children? | ReactNode | A React node that renders custom elements in the tooltip header.
```

File: packages/dx-react-scheduler/docs/reference/appointment-tooltip.md [hunk 7]
```diff
@@ -88,7 +97,9 @@ Describes properties passed to a component that renders the tooltip content.
 Field | Type | Description
 ------|------|------------
 appointmentData? | [AppointmentModel](./scheduler.md#appointmentmodel) | The appointment's displayed metadata.
-children? | ReactNode | A React node used to render the tooltip content.
+formatDate | [FormatterFn](scheduler.md#formatterfn) | A function that formats dates according to the locale.
+recurringIconComponent | ComponentType&lt;object&gt; | A component that renders an icon for recurring appointments inside the AppointmentTooltip.
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/reference/appointment-tooltip.md]:
```suggestion
recurringIconComponent | ComponentType&lt;object&gt; | A component that renders an icon that indicates a recurring appointment.
```

File: packages/dx-react-scheduler/docs/reference/appointment-tooltip.md [hunk 8]
```diff
@@ -88,7 +97,9 @@ Describes properties passed to a component that renders the tooltip content.
...
+children? | ReactNode | A React node used to render the additional part of the tooltip content.
```
Comment (RomanTsukanov) [inline packages/dx-react-scheduler/docs/reference/appointment-tooltip.md]:
```suggestion
children? | ReactNode | A React node that renders custom elements in the tooltip.
```

File: packages/dx-react-scheduler/src/types/editing/appointment-tooltip.types.ts [hunk 9]
```diff
@@ -28,19 +30,41 @@ export namespace AppointmentTooltip {
...
     /** The appointment’s displayed metadata. */
     appointmentData?: AppointmentModel;
-    /** A React node used to render the tooltip header. */
+    /** Specifies the Open button’s visibility. */
+    showOpenButton: boolean;
+    /** Specifies the Close button’s visibility. */
+    showCloseButton: boolean;
+    /** Specifies the Delete button’s visibility. */
+    showDeleteButton: boolean;
+    /** A command button’s identifier list. */
```
Comment (AvicusDelacroix) [inline packages/dx-react-scheduler/src/types/editing/appointment-tooltip.types.ts]: It looks like it should be "The command button's...", but if it's just a comment for internal use, it's ok.

File: packages/dx-react-scheduler/docs/reference/appointment-tooltip.md [hunk 10]
```diff
@@ -64,22 +65,30 @@ showDeleteButton | boolean | Specifies the Delete button's visibility.
 commandButtonIds | Array&lt;string&gt; | A command button's identifier list.
```
Comment (AvicusDelacroix) [inline packages/dx-react-scheduler/docs/reference/appointment-tooltip.md]: "The command button", if you are talking about a specific button.

File: packages/dx-react-scheduler/src/types/editing/appointment-tooltip.types.ts [hunk 11]
```diff
@@ -28,19 +30,41 @@ export namespace AppointmentTooltip {
...
+    commandButtonIds: Array<string>;
+    /** An event raised when the Open button is clicked. The event handler should open the appointment form. */
+    onOpenButtonClick?: () => void;
+    /** An event raised when the Delete button is clicked. The event handler should delete an appointment. */
+    onDeleteButtonClick?: () => void;
+    /** An event that hides the tooltip. */
+    onHide?: () => void;
+    /** A component that renders a command button. */
+    commandButtonComponent: React.ComponentType<AppointmentTooltip.CommandButtonProps>;
+    /** A React node used to render the additional part of the tooltip header. */
```
Comment (AvicusDelacroix) [inline packages/dx-react-scheduler/src/types/editing/appointment-tooltip.types.ts]: There's only one additional part? If there can be different parts, use "an additional part".

File: packages/dx-react-scheduler/src/types/editing/appointment-tooltip.types.ts [hunk 12]
```diff
@@ -28,19 +30,41 @@ export namespace AppointmentTooltip {
...
   /** Describes properties passed to a component that renders the tooltip content. */
   export interface ContentProps {
     /** The appointment’s displayed metadata. */
     appointmentData?: AppointmentModel;
-    /** A React node used to render the tooltip content. */
+    /** A function that formats dates according to the locale. */
+    formatDate: FormatterFn;
+    /** A component that renders an icon for recurring appointments inside the AppointmentTooltip. */
+    recurringIconComponent: React.ComponentType<object>;
+    /** A React node used to render the additional part of the tooltip content. */
```
Comment (AvicusDelacroix) [inline packages/dx-react-scheduler/src/types/editing/appointment-tooltip.types.ts]: Same as the previous comment.

**Comment by boboci9** [comment]
Hi, Is there a way to remove the default tooltipcontent?
I have my own tooltip content designed with different icons based on the appointment type etc. Can disable the default appointmentcontent in this case?
Screen Shot 2020-08-04 at 2 06 35 PM

**Pull request merged**